### PR TITLE
Working version of TFJS test suite

### PR DESCRIPTION
**Pull request by djalova**
## Summary:
* Implemented Javascript versions of `prepost.py` and `handlers.py`.
* Added test script to load model and run a prediction using classes implemented above.

### Current TODOs:
* Look into JS testing frameworks

### How to run:
1. After running the steps in the README, go into the `tensorflowjs` directory and run `python convert.py` to create a tfjs version of the model.
2.  Run `npm install`
3. Run `node test/js/testLocal.js`

### CODE REVIEW

File: tensorflowjs/prePost.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        throw {name: "NotImplementedError", message: "This function needs to be implemented."}
+    }
+}
+
+class ObjectDetectorHandler extends PrePost {
+
+    preprocess(request) {
+    
+        request.processedInputs['image'] = request.rawInputs['image'].map(element => {
+            let base64Encoded = Buffer.from(element).toString('base64');
```
Comment (frreiss) [inline tensorflowjs/prePost.js]: The input image should already be in base64 format when sent to the web service. If you are finding that it is not in that format at this point in the flow, then Node is decoding the base64 string for you.

File: tensorflowjs/prePost.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            const img = new Image();
+            /** 
+             * Note: Setting image src via data url, to avoid waiting for image to load
+             *  See https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Using_images#Embedding_an_image_via_data_URL
+             */
+            img.src = `data:image/jpeg;base64,${base64Encoded}`;
+            const canvas = createCanvas(img.width, img.height);
```
Comment (frreiss) [inline tensorflowjs/prePost.js]: This canvas creation should be in the object constructor if possible.

File: tensorflowjs/prePost.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            return inputTensor;
+            // let base64Encoded = Buffer.from(element).toString('base64');
+            // let imgBinary = new Buffer(base64Encoded, 'base64');
+            // console.log(imgBinary)
+            // return imgBinary;
+        });
+    }
+
+    postprocess(request) {
+        let categories = JSON.parse(fs.readFileSync("test/labels.json"));
```
Comment (frreiss) [inline tensorflowjs/prePost.js]: This file I/O should be in the object constructor.

File: tensorflowjs/test/js/testLocal.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    // console.log(model.executor._outputs)
+    
+    let request = new inferenceRequest.InferenceRequest();
+    let handler = new prepost.ObjectDetectorHandler();
+
+    fetch(SAMPLE_IMAGE_URL)
+        .then(function(response) {
+            return response.arrayBuffer();
+        }).then(function(buffer) {
+             request.rawInputs['image'] = [new Uint8ClampedArray(buffer)];
```
Comment (frreiss) [inline tensorflowjs/test/js/testLocal.js]: This line should be base64-encoding the image.

File: tensorflowjs/prePost.js (hunk 3 above)
Comment (frreiss) [reply tensorflowjs/prePost.js]: Also you should include a comment somewhere in the repo about where you obtained the contents of labels.json

**Review by frreiss — APPROVED** [review]
Looking good so far, some minor suggestions here and there.

**Comment by djalova** [comment]
I updated the PR with @frreiss's suggestions.

There is still a dependency issue since the TFJS converter requires `tensorflow_hub`

**Comment by djalova** [comment]
I made some changes to the environment.
I installed TFJS 0.8.5 with pip using `--no-deps` and installed tensorflow-hub 0.1.1 but running the conversion to TFJS cause it to fail with the following stack trace:
```
Writing weight file web_model/model.json...
Traceback (most recent call last):
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1292, in _do_call
    return fn(*args)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1275, in _run_fn
    self._extend_graph()
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1312, in _extend_graph
    tf_session.ExtendSession(self._session)
tensorflow.python.framework.errors_impl.InvalidArgumentError: No OpKernel was registered to support Op 'DataFormatVecPermute' with these attrs.  Registered devices: [CPU], Registered kernels:
  device='CPU'; T in [DT_INT32]
  device='CPU'; T in [DT_INT64]

	 [[{{node BoxPredictor_0/Shape-0-0-VecPermuteNCHWToNHWC-LayoutOptimizer}} = DataFormatVecPermute[T=DT_INT32, _kernel="host", dst_format="NHWC", src_format="NCHW"](BoxPredictor_0/Shape)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "convert.py", line 22, in <module>
    main()
  File "convert.py", line 18, in main
    convert_to_tfjs(saved_model_path, output_path)
  File "convert.py", line 14, in convert_to_tfjs
    skip_op_check=True)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflowjs/converters/tf_saved_model_conversion.py", line 323, in convert_tf_saved_model
    skip_op_check=skip_op_check, strip_debug_ops=strip_debug_ops)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflowjs/converters/tf_saved_model_conversion.py", line 144, in optimize_graph
    extract_weights(optimized_graph, output_graph, quantization_dtype)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflowjs/converters/tf_saved_model_conversion.py", line 174, in extract_weights
    value = tensor.eval(session=sess)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 711, in eval
    return _eval_using_default_session(self, feed_dict, self.graph, session)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 5155, in _eval_using_default_session
    return session.run(tensors, feed_dict)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 887, in run
    run_metadata_ptr)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1110, in _run
    feed_dict_tensor, options, run_metadata)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1286, in _do_run
    run_metadata)
  File "/Users/djalova/projects/reMAX/Refactored-MAX-Object-Detector/env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1308, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: No OpKernel was registered to support Op 'DataFormatVecPermute' with these attrs.  Registered devices: [CPU], Registered kernels:
  device='CPU'; T in [DT_INT32]
  device='CPU'; T in [DT_INT64]

```

**Comment by djalova** [comment]
I asked @feihugis to help me debug this.
The converter script works on TF 1.13 but not TF1.11.
We found that the TFJS converter script runs its own optimization on the graph and the resulting graph differs between the two versions. In TF 1.11 the graph has a node called `BoxPredictor_0/Shape-0-0-VecPermuteNCHWToNHWC-LayoutOptimizer`, but the TF 1.13 version does not. As part of the conversion process it assigns the GPU device as an attribute to the op, which is not registered as a valid attribute for the operation.

This issue does not exist in TF 1.13 because the operation is not present in the optimized graph.

Thanks for taking a look into this Fei.

**Comment by djalova** [comment]
I bumped the version of TF to 1.13.1 and checked that the non-WML parts of the guide and the TFJS instructions still work.

**Pull request merged**